### PR TITLE
Config: make all compile-time parameters runtime parameters

### LINKED ISSUE #57: Get the Lua path dynamically
At least three files (`gt/main.c`, `config/static.c`, `config/dynamic.c`) hard code the path to the Lua directory. We should find a way of making this dynamic. Since it effects the basic configuration of Gatekeeper -- even before Lua is invoked -- we may need to pass it from the Makefile.

This also relates to issue #14, and might be solved in that patch if it's simple enough.

### PR DESCRIPTION
**Pull request by mengxiang0811**
This pull request closes #57

### CODE REVIEW

**Comment by mengxiang0811** [comment]
This rebase incorporated the latest closed pull requests.

File: include/gatekeeper_config.h [hunk 1]
```diff
@@ -25,66 +25,18 @@
...
- * The maximum cache size can be adjusted in DPDK's .config file: 
- * CONFIG_RTE_MEMPOOL_CACHE_MAX_SIZE.
- */
-#define GATEKEEPER_CACHE_SIZE	(512)
-
 extern const uint16_t LUA_MSG_MAX_LEN;
 
+/* Configuration for the Gatekeeper functional blocks. */
+struct gatekeeper_config {
+	uint8_t  gatekeeper_max_ports;
```
Comment (AltraMayor) [inline include/gatekeeper_config.h]: 1. Instead of defining another global `struct` in Gatekeeper, shouldn't we move the fields of `struct gatekeeper_config` into `struct net_config`? The fields in `struct gatekeeper_config` are even related to the network.

2. The constants that this patch is moving to the configuration files were created before we have the current structure of configuration files. So, isn't it possible for at least some of these constants to be dynamically calculated as we do with lcores and number of queues for example?

File: config/dynamic.c [hunk 2]
```diff
@@ -514,17 +512,33 @@ run_dynamic_config(struct gk_config *gk_conf, struct gt_config *gt_conf,
 			"dyn_cfg: Failed to unlink(%s) - (%s)\n",
 			dy_conf->server_path, strerror(errno));
 		ret = -1;
-		goto free_path;
+		goto free_server_path;
 	}
 
+	dy_conf->lua_dy_base_dir = rte_malloc(
+		"lua_dy_base_dir", strlen(lua_dy_base_dir) + 1, 0);
```
Comment (AltraMayor) [inline config/dynamic.c]: 1. Rename `gt_strdup()` to `rte_strdup()`, move it to a proper location, and apply it here. Notice that `gt_strdup()` was introduced by a recently merged pull request.

2. I can't comment there, but apply `rte_strdup()` on `dy_conf->server_path` as well.

File: config/dynamic.c [hunk 3]
```diff
@@ -514,17 +512,33 @@ run_dynamic_config(struct gk_config *gk_conf, struct gt_config *gt_conf,
...
+	if (dy_conf->lua_dy_base_dir == NULL) {
+		ret = -1;
+		goto free_server_path;
+	}
+	strcpy(dy_conf->lua_dy_base_dir, lua_dy_base_dir);
+
+	dy_conf->dynamic_config_file = rte_malloc(
+		"dynamic_config_file", strlen(dynamic_config_file) + 1, 0);
```
Comment (AltraMayor) [inline config/dynamic.c]: 1. Apply `rte_strdup()` here as well.

2. You forgot to update `cleanup_dy()` for `dy_conf->lua_dy_base_dir` and `dy_conf->dynamic_config_file`.

File: include/gatekeeper_config.h [hunk 4]
```diff
@@ -62,6 +62,12 @@ struct dynamic_config {
 
 	/* Specify the receiving timeouts until reporting an error. */
 	struct timeval   rcv_time_out;
+
+	/* The directory for Lua files of dynamic configuration. */
+	char             *lua_dy_base_dir;
```
Comment (AltraMayor) [inline include/gatekeeper_config.h]: The line above and the lines around it are wrongly indented.

File: config/dynamic.c [hunk 5]
```diff
@@ -590,7 +604,15 @@ run_dynamic_config(struct gk_config *gk_conf, struct gt_config *gt_conf,
...
-free_path:
+free_dynamic_config_file:
+	rte_free(dy_conf->dynamic_config_file);
+	dy_conf->dynamic_config_file = NULL;
+
+free_dy_lua_base_dir:
+	rte_free(dy_conf->lua_dy_base_dir);
+	dy_conf->lua_dy_base_dir = NULL;
+
+free_server_path:
```
Comment (AltraMayor) [inline config/dynamic.c]: Let's drop the blank lines before the labels in this chain of error recovery.

File: Makefile [hunk 6]
```diff
@@ -48,7 +48,8 @@ SRCS-y += lib/mailbox.c lib/net.c lib/flow.c lib/ipip.c \
 
 LDLIBS += $(LDIR) -Bstatic -lluajit-5.1 -Bdynamic -lm -lmnl
 CFLAGS += $(WERROR_FLAGS) -I${GATEKEEPER}/include -I/usr/local/include/luajit-2.0/
-EXTRA_CFLAGS += -O3 -g -Wfatal-errors -DALLOW_EXPERIMENTAL_API
+EXTRA_CFLAGS += -O3 -g -Wfatal-errors -DALLOW_EXPERIMENTAL_API \
+	-DLUA_BASE_DIR='"./lua"' -DGATEKEEPER_CONFIG_FILE='"gatekeeper_config.lua"'
```
[the remark this answers is not in the document]
Comment (mengxiang0811) [reply Makefile]: 1. No. Otherwise, we will get errors: ``/home/qiaobin/gatekeeper/config/static.c: In function ‘config_gatekeeper’:``
``<command-line>:0:14: error: expected expression before ‘.’ token``
`` note: in expansion of macro ‘LUA_BASE_DIR’
    "%s/%s", LUA_BASE_DIR, GATEKEEPER_CONFIG_FILE);``

File: include/gatekeeper_config.h (hunk 4 above)
Comment (mengxiang0811) [reply include/gatekeeper_config.h]: It seems the indent is correct.

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (AltraMayor) [reply Makefile]: 1. Thank you for trying it out.

File: include/gatekeeper_config.h (hunk 4 above)
Comment (AltraMayor) [reply include/gatekeeper_config.h]: It's showing right to me now. I guess it a GitHub bug.

**Comment by mengxiang0811** [comment]
This rebase incorporated the latest suggestions except for accepting parameters. I am working on this suggestion.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

**Comment by AltraMayor** [comment]
About patch **config: get the Lua path via application parameters**. Given that the whole process of having application parameters consists of skipping EAL parameters and do regular parameter processing, why did you use a different library to process parameters? I'm just thinking of the maintenance cost here since all of our code already use argp.

**Comment by mengxiang0811** [comment]
I have checked several DPDK examples, all are using separate library to process the parameters. For example the [L3FWD](https://github.com/cjdoucette/dpdk/blob/46a945c1a3bf5904cb88c28bf1e8c441e0536641/examples/l3fwd/main.c#L816).

**Comment by AltraMayor** [comment]
I understood from the code that DPDK expects us to use a library to process the parameters of our application. My concern is to use a parameter library other than argp that we've been using in all our projects.

**Comment by mengxiang0811** [comment]
I see your point. I will stick to the argp.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: main/main.c [hunk 7]
```diff
@@ -142,10 +196,29 @@ run_signal_handler(void)
 int
 main(int argc, char **argv)
 {
+	struct args args = {
+		.lua_base_dir = NULL,
```
Comment (AltraMayor) [inline main/main.c]: 1. Replace `NULL` with `LUA_BASE_DIR` since this is a default value.

2. Add comment `/* Defaults. */`  just before the line above.

File: main/main.c [hunk 8]
```diff
@@ -142,10 +196,29 @@ run_signal_handler(void)
...
+		.gatekeeper_config_file = NULL,
```
Comment (AltraMayor) [inline main/main.c]: Replace `NULL` with `GATEKEEPER_CONFIG_FILE` since this is a default value.

File: main/main.c [hunk 9]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
...
+		args->lua_base_dir = arg;
+		break;
+
+	case 'f':
+		args->gatekeeper_config_file = arg;
+		break;
+
+	case ARGP_KEY_INIT:
+		args->lua_base_dir = NULL;
+		args->gatekeeper_config_file = NULL;
```
Comment (AltraMayor) [inline main/main.c]: Once `args->lua_base_dir` and `args->gatekeeper_config_file` have default values, we can drop the two assignments above.

File: config/static.c [hunk 10]
```diff
@@ -246,9 +246,12 @@ config_gatekeeper(void)
 	int ret;
 	char lua_entry_path[128];
 	lua_State *lua_state;
+	const char *base = lua_base_dir == NULL ? LUA_BASE_DIR : lua_base_dir;
+	const char *file = gatekeeper_config_file == NULL ?
+		GATEKEEPER_CONFIG_FILE : gatekeeper_config_file;
```
Comment (AltraMayor) [inline config/static.c]: Thanks to the default values, we can use variables `lua_base_dir` and `gatekeeper_config_file`  directly.

File: main/main.c [hunk 11]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
 uint64_t cycles_per_ms;
 uint64_t picosec_per_cycle;
 
+char *lua_base_dir = NULL;
+char *gatekeeper_config_file = NULL;
```
Comment (AltraMayor) [inline main/main.c]: 1. Can't we add `const`  before `char` for both variables? If so, remember to update the header file `include/gatekeeper_main.h`.

2. Since they are **global** variables, they are automatically initialized to `NULL`, so drop the initialization.

File: main/main.c [hunk 12]
```diff
@@ -142,10 +196,29 @@ run_signal_handler(void)
...
+
+	/* Parse Gatekeeper arguments (after the EAL ones). */
+	ret = argp_parse(&argp, argc, argv, 0, NULL, &args);
+	if (ret != 0)
+		rte_exit(EXIT_FAILURE, "Invalid Gatekeeper parameters\n");
+
+	if (args.lua_base_dir)
+		lua_base_dir = rte_strdup("lua_base_dir", args.lua_base_dir);
+	if (args.gatekeeper_config_file)
+		gatekeeper_config_file = rte_strdup("gatekeeper_config_file",
```
Comment (AltraMayor) [inline main/main.c]: Given that `args.lua_base_dir` and `args.gatekeeper_config_file` points to memory that will only go away once Gatekeeper finishes, we can directly assign them to the global variables.

File: main/main.c [hunk 13]
```diff
@@ -186,5 +259,9 @@ main(int argc, char **argv)
 net:
 	gatekeeper_free_network();
 out:
+	rte_free(lua_base_dir);
+	lua_base_dir = NULL;
+	rte_free(gatekeeper_config_file);
+	gatekeeper_config_file = NULL;
```
Comment (AltraMayor) [inline main/main.c]: Once the global variables are initialized differently (see another comment), we no longer need to deallocate memory here.

File: config/static.c [hunk 14]
```diff
@@ -246,9 +246,12 @@ config_gatekeeper(void)
...
 
 	ret = snprintf(lua_entry_path, sizeof(lua_entry_path), \
-			"%s/%s", LUA_BASE_DIR, GATEKEEPER_CONFIG_FILE);
+			"%s/%s", base, file);
 	RTE_VERIFY(ret > 0 && ret < (int)sizeof(lua_entry_path));
 
 	lua_state = luaL_newstate();
```
Comment (AltraMayor) [inline config/static.c]: I can't comment at line `set_lua_path(lua_state, LUA_BASE_DIR);` below, but you forgot to edit it.

File: main/main.c [hunk 15]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
...
+	{"lua-base-dir", 'd', "DIR", 0,
+		"Base directory DIR for Gatekeeper Lua files", 0},
+	{"gatekeeper-config-file", 'f', "FILE", 0,
+		"Lua configuration FILE to initialize Gatekeeper", 0},
+	{ 0 }
+};
+
+struct args {
+	char *lua_base_dir;
+	char *gatekeeper_config_file;
```
Comment (AltraMayor) [inline main/main.c]: Add `const` before `char` for both fields.

File: main/main.c [hunk 16]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
 uint64_t cycles_per_ms;
 uint64_t picosec_per_cycle;
 
+char *lua_base_dir = NULL;
+char *gatekeeper_config_file = NULL;
+
+/* Argp's global variables. */
+const char *argp_program_version = "gatekeeper 1.0";
```
Comment (AltraMayor) [inline main/main.c]: Capitalize the `G` in `gatekeeper`.

File: main/main.c [hunk 17]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
...
+
+/* Arguments. */
+static char adoc[] = "<LUA_BASE_DIR> <GATEKEEPER_CONFIG_FILE>";
+
+static char doc[] = "gatekeeper -- the first open source "
+	"DDoS protection system";
```
Comment (AltraMayor) [inline main/main.c]: Capitalize the `G` in `gatekeeper`.

File: main/main.c [hunk 18]
```diff
@@ -45,6 +46,59 @@ uint64_t cycles_per_sec;
...
 uint64_t picosec_per_cycle;
 
+char *lua_base_dir = NULL;
+char *gatekeeper_config_file = NULL;
+
+/* Argp's global variables. */
+const char *argp_program_version = "gatekeeper 1.0";
+
+/* Arguments. */
+static char adoc[] = "<LUA_BASE_DIR> <GATEKEEPER_CONFIG_FILE>";
```
Comment (AltraMayor) [inline main/main.c]: Variable `adoc` should be `""` (i.e. the empty string) because Gatekeeper requires no arguments. Gatekeeper has options, though.

File: config/dynamic.c [hunk 19]
```diff
@@ -475,9 +477,24 @@ set_dyc_timeout(unsigned int sec,
 	dy_conf->rcv_time_out.tv_usec = usec;
 }
 
+char *
+rte_strdup(const char *type, const char *s)
+{
+	char *res = rte_malloc(type, strlen(s) + 1, 0);
+	if (res == NULL) {
+		RTE_LOG(ERR, MALLOC, "gt: out of memory for duplicating the string: %s\n", s);
+		goto out;
```
[the remark this answers is not in the document]
Comment (AltraMayor) [reply config/dynamic.c]: 3. Add `unlikely()` to the `if` test: `if (unlikely(res == NULL))`

File: config/dynamic.c [hunk 20]
```diff
@@ -475,9 +477,24 @@ set_dyc_timeout(unsigned int sec,
 	dy_conf->rcv_time_out.tv_usec = usec;
 }
 
+char *
+rte_strdup(const char *type, const char *s)
```
[the remark this answers is not in the document]
Comment (mengxiang0811) [reply config/dynamic.c]: In the ``lib/launch.c`` file, we already had one similar function named ``rte_strdup(const char *s)``.

**Comment by mengxiang0811** [comment]
This rebase is ready for anther review.

File: include/gatekeeper_config.h (hunk 1 above)
Comment (AltraMayor) [reply include/gatekeeper_config.h]: 2. It doesn't seem to be the case. For example, `GATEKEEPER_MAX_PORTS` is used in `gatekeeper_init_network()` only to check that the number of ports doesn't go above that limit. So, if we remove that constant from Gatekeeper, Gatekeeper should still work fine. The real need for `GATEKEEPER_MAX_PORTS` is to find `GATEKEEPER_MBUF_SIZE` and `GATEKEEPER_CACHE_SIZE` as their comments point out. Moreover, the number of needed ports can be easily computed from variables `front_ports` and `back_ports` in `lua/net.lua`.

Although we need to work further on patch **config: make net compile-time parameters runtime parameters**, all the other patches in this pull request are ready for merge. So I suggest to move patch **config: make net compile-time parameters runtime parameters** into a new pull request. Once you do this move, remember to edit the description of the current pull request to only close issue #57 since the new pull request will close issue #14.

**Comment by mengxiang0811** [comment]
This rebase dropped the patch **config: make net compile-time parameters runtime parameters**

File: main/main.c [hunk 21]
```diff
@@ -142,10 +191,27 @@ run_signal_handler(void)
...
+	argc -= ret;
+	argv += ret;
+
+	/* Parse Gatekeeper arguments (after the EAL ones). */
+	ret = argp_parse(&argp, argc, argv, 0, NULL, &args);
+	if (ret != 0)
+		rte_exit(EXIT_FAILURE, "Invalid Gatekeeper parameters\n");
+
+	lua_base_dir = args.lua_base_dir;
+	gatekeeper_config_file = args.gatekeeper_config_file;
```
Comment (AltraMayor) [inline main/main.c]: Let's pass `args.lua_base_dir` and `args.gatekeeper_config_file` as parameters of `config_gatekeeper()` and get rid of the global variables `lua_base_dir` and `gatekeeper_config_file`.

I know that I've said that this patch was ready, but I couldn't resist this cute improvement once I saw it.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.
